pkg/controller/pod.go

```diff
@@ -202,15 +214,93 @@ func (c *Controller) syncPod(ctx context.Context, pod *corev1.Pod) error {
 		}
 	}
 
-	klog.V(4).Infof("Registering instance %s under service %+v", pod.Name, cloudmapConfig)
+	cloudMapServiceKey := *(awssdk.String(*cloudmapConfig.NamespaceName)) + "-" +
+		*(awssdk.String(*cloudmapConfig.ServiceName))
+
+	existingItem, exists, _ := c.cloudMapInstanceCache.Get(&CloudMapInstanceCacheItem{
+		key: cloudMapServiceKey,
+	})
+
+	var serviceInstanceSummary map[string]bool
+
+	if exists {
+		if healthy, ok := existingItem.(*CloudMapInstanceCacheItem).InstanceSummary[pod.Status.PodIP]; ok {
+			//Right now we're not logging pod/instance health info to CloudMap. So, if it is in cache then it is
+			//considered healthy. *TODO - v1.0* */
+			klog.V(4).Info("Pod Instance already registered. Nothing to do for this pod; Healthy?: ", healthy)
+			return nil
+		}
+		serviceInstanceSummary = existingItem.(*CloudMapInstanceCacheItem).InstanceSummary
+		if serviceInstanceSummary == nil { serviceInstanceSummary = make(map[string]bool) }
+		serviceInstanceSummary[pod.Status.PodIP] = true
+	} else {
+		//Retrieve CloudMap instances for this service
+		serviceInstanceSummary := make(map[string]bool)
+		registered := false
+
+		appmeshCloudMapConfig := &appmesh.AwsCloudMapServiceDiscovery{
+			NamespaceName: awssdk.String(*cloudmapConfig.NamespaceName),
+			ServiceName:   awssdk.String(*cloudmapConfig.ServiceName),
+		}
+
+		if serviceInstances, _ := c.getServiceInstancesFromCloudMap(ctx, appmeshCloudMapConfig); len(serviceInstances) > 0 {
+			for _, instance := range serviceInstances {
+				podName := awssdk.StringValue(instance.Attributes[ctrlaws.AttrK8sPod])
+				podNamespace := awssdk.StringValue(instance.Attributes[ctrlaws.AttrK8sNamespace])
+				instanceID := awssdk.StringValue(instance.Id)
+				serviceName := awssdk.StringValue(instance.Attributes[ctrlaws.AttrK8sApp])
+
+				if instanceID == pod.Status.PodIP {
+					klog.V(4).Info("Instance already registered for ServiceName: ", podName, serviceName)
+					registered = true
+				}
+
+				cloudMapServiceKey = podNamespace + ".pvt.aws.local" + "-" + serviceName
+				serviceInstanceSummary[instanceID] = true
+			}
+
+			serviceItem := &CloudMapInstanceCacheItem{
+				key:             cloudMapServiceKey,
+				InstanceSummary: serviceInstanceSummary,
+			}
+			_ = c.cloudMapInstanceCache.Add(serviceItem)
+
+			if registered {
+				klog.V(4).Info("Instance already registered. Cache Refreshed. Move on")
+				return nil
+			}
+		}
+	}
+
```

Comment:
seems this cache can only reduce `RegisterInstance` call if the cache for service is not present?
e.g. if the pod get updated twice(maybe some annotation or condition updates or just regular resync) , RegisterInstance will still be called even cache for the service contains it.

Reply:
No, it will not make the registerinstance call in the scenario you described as well